### ISSUE TITLE
Module detection broken in monorepos

### ISSUE THREAD

**Issue by nfantone** [weiran-zsd/eslint-plugin-node#37]
[This line](https://github.com/weiran-zsd/eslint-plugin-node/blob/a76638206cb6a92c2f8777128a0e9ca5c48aec12/lib/configs/recommended.js#L9) is always fetching the `"type"` value from the root level `package.json` instead of the nearest one, which produces false positives on monorepos with a mix of `"commonjs"` and `"module"` packages.

Additionally, `eslint` already provides settings for determining which type of package the current project is: `parserOptions.sourceType`. IMHO, we should be using _that_ instead of leveraging `package.json` properties. As of right now, even if I have `"sourceType"` set to `"script"`, this plugin may produce errors out of `module.exports` and `require` usages — which should not be the case.

**Comment by aladdin-add** [comment]
your own config should have overwrite the plugin's `recommended` preset. so the parserOptions is working as you expected, the eslint errors  comes from:
https://github.com/weiran-zsd/eslint-plugin-node/blob/a76638206cb6a92c2f8777128a0e9ca5c48aec12/lib/configs/recommended-module.js#L6-L12


you are able to use the glob config, something like:

```js
  "overrides": [
    {
      "files": ["packageA/"],
      "extends": ["plugin:n/recommended-module"]
    },
    {
      "files": ["packageB/"],
      "extends": ["plugin:n/recommended-script"]
    },
  ]
```

**Comment by nfantone** [comment]
Hi @aladdin-add, thanks for your reply.

Not sure I got where you're coming from, though.

> your own config should have overwrite the plugin's recommended preset.

What do you mean? At first glance, that doesn't seem to be the case for me. I'm not purposely overriding anything in my config. This is what I have:

```js
{
  extends: [
    'standard',
    'plugin:n/recommended',
    'plugin:jest/recommended',
    'plugin:prettier/recommended',
  ],
  parserOptions: {
    sourceType: 'script',
  },
}
```

> the `parserOptions` is working as you expected

I might be wrong here, but I don't think it is. And I say that because:

* If I remove the plugin from my `extends` array, without doing anything else, all linting errors go away.
* If I change `"type"` to `"commonjs"` in my **root** `package.json` (regardless of what `parserOptions` holds), once again, `eslint` errors are gone.

**Comment by aladdin-add** [comment]
the preset was doing 2 things:
1. set the `parserOptions: {"sourceType": "script"/"module"}` based on your package.json
2. turn on/off cjs globals

In your case, you only specified your own `parserOptions: {sourceType: 'script',}`, but cjs globals are still disallowed.

If your project is a monorepo, and mixed esm/cjs. I strongly recommend you not to use the `plugin:n/recommended`, but using the `plugin:n/recommended-module` and `plugin:n/recommended-script`. 

just like my last comment, you can see the [glob config docs](https://eslint.org/docs/latest/user-guide/configuring/configuration-files#configuration-based-on-glob-patterns).

**Comment by nfantone** [comment]
I see. Thanks for sharing. Will give the `n/recommended-module` and `n/recommended-script` presets a go. 

> In your case, you only specified your own `parserOptions: {sourceType: 'script',}` , but cjs globals are still disallowed.

Which, in my view, is unexpected and, frankly, wrong. The plugin should not do that and tt leads to inconsistencies. If my `sourceType` is `'script'`, CJS globals would need to be allowed.

**Comment by blephy** [comment]
Hey,

Another solution would to have an eslint config by packages, extended from a base config, but with a root: true.

**Comment by scagood** [comment]
> Hey,
> 
> Another solution would to have an eslint config by packages, extended from a base config, but with a root: true. This way, no need to maintain a set of overrides, but you will be forced to run eslint independently by packages.

This is exactly the same thing I do in my monorepos. Each sub package has its own eslint config, which references a global config.

<details>
<summary>In that global config I do a similar detection to the one in the base recommended config</summary>

![image](https://github.com/eslint-community/eslint-plugin-n/assets/2230835/f917948d-05b1-4e02-a2bd-2b9bbb9bf180)

</details>

The `root` variable is actually settable directly from the each of the package's configs.
```js
import { createConfig } from '@scagood/eslint';

export default createConfig(import.meta.url);
```
